fabtests/multinode/src/timing.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2022 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <string.h>
+#include <limits.h>
+#include <shared.h>
+#include <timing.h>
+#include <core.h>
+
+extern struct pm_job_info pm_job;
+
+int multi_timer_create(struct multi_timer **timer, int rank)
+{
+	*timer = malloc(sizeof(*timer));
+
+	(*timer)->rank = rank;
+	(*timer)->start = -1;
+	(*timer)->end = -1;
+
+	return 0;
+}
+
+void multi_timer_start(struct multi_timer *timer)
+{
+	if (timer && timer->start == -1)
+		timer->start = ft_gettime_ns();
+}
+
+void multi_timer_stop(struct multi_timer *timer)
+{
+	if (timer && timer->start != -1 && timer->end == -1)
+		timer->end = ft_gettime_ns();
+}
+
+int multi_timer_analyze(struct multi_timer **timers, int timer_count)
+{
+	int i, j, ret = 0;
+	int iterations = timer_count / pm_job.num_ranks;
+	double total_timers, total_min, total_max, total_sum_time, total_duration;
+	long *min = malloc(sizeof(*min) * iterations);
+	long *max = malloc(sizeof(*max) * iterations);
+	long *first_start = malloc(sizeof(*first_start) * iterations);
+	long *last_end = malloc(sizeof(*last_end) * iterations);
+	double *sum_time = malloc(sizeof(*sum_time) * iterations);
```

Comment:
All malloc calls would be safer as calloc, and are more natural for calloc anyway.